preconvert to PIL.Image; optionally preconvert to torch.Tensor

bkj opened the pull request:
- For some datasets (cifar, svhn, stl10, mnist, fashionMNIST), conversion from numpy arrays to PIL.Image is done each time an image is requested.  This pull preconverts to PIL.Image, which gives a nontrivial speedup (see below code)
- Further, if you don't need to do any data augmentation that operates on the PIL.Image, you can preconvert to torch.Tensor.  This can give a quite a large speedup (see code).

For some applications, this can make a big difference.  Eg, training a relatively simple CNN for MNIST, you get the following speedups (~25% by precomputing PIL.Image, ~50% by precomputing torch.Tensors)

```
# current torchvision implementation
{'epoch': 0, 'train_score': 0.8918333333333334, 'val_score': 0.9763333333333334, 'test_score': 0.9795, 'time': 6.856365919113159}
{'epoch': 1, 'train_score': 0.967462962962963, 'val_score': 0.9855, 'test_score': 0.986, 'time': 5.420516490936279}
{'epoch': 2, 'train_score': 0.9761296296296297, 'val_score': 0.9855, 'test_score': 0.988, 'time': 5.8416588306427}
{'epoch': 3, 'train_score': 0.9797407407407407, 'val_score': 0.9886666666666667, 'test_score': 0.9898, 'time': 5.751266002655029}

# precompute PIL.Image (automatic in new code)
{'epoch': 0, 'train_score': 0.8918518518518519, 'val_score': 0.9756666666666667, 'test_score': 0.9798, 'time': 4.791485548019409}
{'epoch': 1, 'train_score': 0.9675, 'val_score': 0.9835, 'test_score': 0.9862, 'time': 3.9894304275512695}
{'epoch': 2, 'train_score': 0.9757222222222223, 'val_score': 0.9853333333333333, 'test_score': 0.9887, 'time': 3.690760850906372}
{'epoch': 3, 'train_score': 0.9798333333333333, 'val_score': 0.9878333333333333, 'test_score': 0.99, 'time': 3.920809030532837}

# precompute torch.Tensor (using pretensor flag in new code)
{'epoch': 0, 'train_score': 0.891962962962963, 'val_score': 0.9756666666666667, 'test_score': 0.979, 'time': 3.3365728855133057}
{'epoch': 1, 'train_score': 0.9677037037037037, 'val_score': 0.9825, 'test_score': 0.9846, 'time': 2.4013609886169434}
{'epoch': 2, 'train_score': 0.9757777777777777, 'val_score': 0.9851666666666666, 'test_score': 0.9885, 'time': 2.367426633834839}
{'epoch': 3, 'train_score': 0.9804444444444445, 'val_score': 0.987, 'test_score': 0.9907, 'time': 2.481478214263916}
```


Script illustrating relative speed of loaders is [here](https://gist.github.com/bkj/e76a386c49064bdf4b884cf4a8567e66)

soumith commented:
i looked through the PR and code-wise it looks fine.

However, a dataset like SVHN living entirely in memory will be pretty massive, right?

bkj commented:
Ah yeah -- good point.  I'll check memory usage on all of these tomorrow and revert anything that looks like it's going to be too big of a drain.

fmassa commented on torchvision/datasets/folder.py:
```diff
@@ -16,6 +16,9 @@ def is_image_file(filename):
     Returns:
         bool: True if the filename ends with a known image extension
     """
+    if force:
```
is this change necessary? Can you remove this please?

fmassa commented on torchvision/__init__.py:
```diff
@@ -3,7 +3,7 @@
 from torchvision import utils
 
-__version__ = '0.2.0'
+__version__ = '0.2.0.bkj'
```
Hum, I'm not sure we should be manually modifying the `__version__`, ideally we should use the commit hashes, but that's for another story. Can you maybe revert this?

bkj commented:
OK made those changes:
  - removed `force` flag and version change (those were pushed by accident, sorry about that)
  - removed `pretensor` option, so that datasets always return PIL.Image
  - checked memory usage -- for the largest dataset (SVHN unlabeled), the whole thing is already being loaded into memory.  The way the PR was written, two copies of the data were kept in memory unnecessarily (eg `self.data` and `self._data`).  I've fixed that now.

fmassa commented on torchvision/datasets/cifar.py:
```diff
@@ -11,6 +11,7 @@
 
 import torch.utils.data as data
 from .utils import download_url, check_integrity
+from ..transforms import functional as F
```
I believe this is not needed anymore?

fmassa commented on torchvision/datasets/cifar.py:
```diff
@@ -46,9 +47,7 @@ class CIFAR10(data.Dataset):
-    def __init__(self, root, train=True,
-                 transform=None, target_transform=None,
-                 download=False):
+    def __init__(self, root, train=True, transform=None, target_transform=None, download=False):
```
I think this line might be too long and lint is complaining. Can you make it as before?

fmassa commented on torchvision/datasets/cifar.py:
```diff
@@ -83,6 +82,8 @@ def __init__(self, root, train=True,
             self.train_data = self.train_data.reshape((50000, 3, 32, 32))
             self.train_data = self.train_data.transpose((0, 2, 3, 1))  # convert to HWC
+
+            self.train_data = [Image.fromarray(self.train_data[i]) for i in range(self.train_data.shape[0])]
```
I believe this can be simplified as follows
```python
self.train_data = [Image.fromarray(img) for img in self.train_data]
```
Also the line is probably too long and lint is complaining

fmassa commented on torchvision/datasets/cifar.py:
```diff
@@ -100,6 +101,8 @@ def __init__(self, root, train=True,
             self.test_data = self.test_data.reshape((10000, 3, 32, 32))
             self.test_data = self.test_data.transpose((0, 2, 3, 1))  # convert to HWC
 
+            self.test_data = [Image.fromarray(self.test_data[i]) for i in range(self.test_data.shape[0])]
```
Same here

fmassa commented on torchvision/datasets/mnist.py:
```diff
@@ -7,6 +7,7 @@
 import numpy as np
 import torch
 import codecs
+from ..transforms import functional as F
```
This can be removed

fmassa commented on torchvision/datasets/mnist.py:
```diff
@@ -52,10 +53,16 @@ def __init__(self, root, train=True, transform=None, target_transform=None, down
             self.train_data, self.train_labels = torch.load(
                 os.path.join(self.root, self.processed_folder, self.training_file))
+
+            self.train_data = [Image.fromarray(self.train_data[i].numpy(), mode='L')
```
Same here, this can be simplified

fmassa commented on torchvision/datasets/stl10.py:
```diff
@@ -7,6 +7,7 @@
 import numpy as np
 import sys
 from .cifar import CIFAR10
+from ..transforms import functional as F
```
This can be removed

fmassa commented on torchvision/datasets/stl10.py:
```diff
@@ -82,6 +83,8 @@ def __init__(self, root, split='train',
             with open(class_file) as f:
                 self.classes = f.read().splitlines()
 
+        self.data = [Image.fromarray(np.transpose(self.data[i], (1, 2, 0))) for i in range(self.data.shape[0])]
```
This can be simplified as well

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -5,6 +5,7 @@
 import os.path
 import numpy as np
 from .utils import download_url, check_integrity
+from ..transforms import functional as F
```
Please remove

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -81,6 +82,8 @@ def __init__(self, root, split='train',
         np.place(self.labels, self.labels == 10, 0)
         self.data = np.transpose(self.data, (3, 2, 0, 1))
 
+        self.data = [Image.fromarray(np.transpose(self.data[i], (1, 2, 0))) for i in range(self.data.shape[0])]
```
Simplify?
